storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java

```diff
@@ -252,26 +257,43 @@ public SafeFuture<Optional<BeaconState>> getStateAtSlotExact(final UInt64 slot)
   }
 
   public SafeFuture<Optional<BeaconState>> getStateForBlockProduction(
-      final UInt64 slot, final boolean isForkChoiceLateBlockReorgEnabled) {
+      final UInt64 slot,
+      final boolean isForkChoiceLateBlockReorgEnabled,
+      final Runnable onLateBlockPreparationCompleted) {
     if (!isForkChoiceLateBlockReorgEnabled) {
       return getStateAtSlotExact(slot);
     }
-    final Optional<Bytes32> headRoot = getBestBlockRoot();
-    if (headRoot.isEmpty()) {
+    final Optional<ChainHead> chainHead = getChainHead();
+    if (chainHead.isEmpty()) {
       return getStateAtSlotExact(slot);
     }
-    final Bytes32 root = recentChainData.getProposerHead(headRoot.get(), slot);
-    if (root.equals(headRoot.get())) {
+    final Bytes32 headRoot = chainHead.get().getRoot();
+
+    final Bytes32 proposerHeadRoot = recentChainData.getProposerHead(headRoot, slot);
+    if (proposerHeadRoot.equals(headRoot)) {
       return getStateAtSlotExact(slot);
     }
     // otherwise we're looking for the parent slot
-    return getStateByBlockRoot(root)
+    return getStateByBlockRoot(proposerHeadRoot)
         .thenCompose(
             maybeState ->
                 maybeState
                     .map(
-                        beaconState ->
-                            SafeFuture.completedFuture(regenerateBeaconState(beaconState, slot)))
+                        beaconState -> {
+                          // let's run preparation and state generation in parallel
+                          final SafeFuture<Void> preparationCompletion =
+                              lateBlockReorgPreparationHandler
+                                  .onLateBlockReorgPreparation(
+                                      recentChainData
+                                          .getSlotForBlockRoot(proposerHeadRoot)
+                                          .orElseThrow(),
+                                      headRoot)
+                                  .thenPeek(__ -> onLateBlockPreparationCompleted.run());
+                          final Optional<BeaconState> state =
+                              regenerateBeaconState(beaconState, slot);
+
+                          return preparationCompletion.thenApply(__ -> state);
```

Comment:
Potentially, we could have an `orTimeout()` for LBR preparation in the future. I am thinking if there is a case in which we are better off proposing on top of the weak head than risking missing the LBR because preparation took too long. wdyt?